File: src/DesignPatterns/src/C++/001-SOLID/004-LSP/LSPParserApp/Parser.cpp
#include <iostream>
#include <cctype>
#include "Parser.hpp"

using namespace std;

void Parser::SetSource(ISource *src)
{
	//...
	m_src = src;
}

void Parser::DoParse()
{
	auto count = 0;
	int ch;

	while ((ch = m_src->GetNextChar()) != -1)
		if (isspace(ch))
			++count;

	cout << "Count:" << count << endl;
}

File: src/DesignPatterns/src/Java/001-SOLID/003-LSP/ParserApp/src/org/csystem/app/Parser.java
package org.csystem.app;

public class Parser {
    private ISource m_source;

    public Parser(ISource source)
    {
        m_source = source;
    }

    public ISource getSource()
    {
        return m_source;
    }

    public void setSource(ISource source)
    {
        m_source = source;
    }

    public void doParse()
    {
        int count = 0;

        int ch;

        while ((ch = m_source.getChar()) != -1)
            if (Character.isWhitespace((char)ch))
                ++count;

        System.out.printf("count:%d%n", count);
    }
}
